src/Tools/dotnet-monitor/Auth/ApiKeyAuthenticationOptionsChangeTokenSource.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using System;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    /// <summary>
+    /// Notifies that ApiKeyAuthenticationOptions changes when ApiAuthenticationOptions changes.
+    /// </summary>
+    internal sealed class ApiKeyAuthenticationOptionsChangeTokenSource :
+        IOptionsChangeTokenSource<ApiKeyAuthenticationOptions>,
+        IDisposable
+    {
+        private readonly IOptionsMonitor<ApiAuthenticationOptions> _optionsMonitor;
+        private readonly IDisposable _changeRegistration;
+
+        private ConfigurationReloadToken _reloadToken = new ConfigurationReloadToken();
+
+        public ApiKeyAuthenticationOptionsChangeTokenSource(
+            IOptionsMonitor<ApiAuthenticationOptions> optionsMonitor)
+        {
+            _optionsMonitor = optionsMonitor;
+            _changeRegistration = _optionsMonitor.OnChange(OnReload);
+        }
+
+        public string Name => Options.DefaultName;
+
+        public IChangeToken GetChangeToken()
+        {
+            return _reloadToken;
+        }
+
+        public void Dispose()
+        {
+            _changeRegistration.Dispose();
+        }
+
+        private void OnReload(ApiAuthenticationOptions options)
+        {
+            Interlocked.Exchange(ref _reloadToken, new ConfigurationReloadToken()).OnReload();
```

Comment:
So we can get multiple OnReload's on different threads? Does that mean we can have multiple OnChange operations happening at the same time?

Reply:
That may be possible, but this is to make sure that multiple callers of GetChangeToken get the same token if they happen to call in when the options monitor is signaling the change. Implementation is basically the same as https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs

Reply:
I'm wondering if Interlocked was just used for convenience of assigning a field while returning the previous field. It would be odd for any of this to be multi-threaded.

Reply:
You're probably correct for any given chain of updates. I do see that it is a different thread for each change, but the entire change starting from the KeyPerFileConfigurationProvider change callback all the way to when the observer wants to log the validation errors is all a single-threaded single execution context.